server/src/main/java/org/elasticsearch/cluster/ClusterFeatures.java

```diff
@@ -79,28 +82,74 @@ public Map<String, Set<String>> nodeFeatures() {
         return nodeFeatures;
     }
 
-    /**
-     * The features in all nodes in the cluster.
-     * <p>
-     * NOTE: This should not be used directly.
-     * Please use {@link org.elasticsearch.features.FeatureService#clusterHasFeature} instead.
-     */
-    public Set<String> allNodeFeatures() {
+    private Set<String> allNodeFeatures() {
         if (allNodeFeatures == null) {
             allNodeFeatures = Set.copyOf(calculateAllNodeFeatures(nodeFeatures.values()));
         }
         return allNodeFeatures;
     }
 
+    /**
+     * Returns {@code true} if {@code node} can have assumed features.
+     * <p>
+     * This is true if it is of the next major version, or it is running on serverless.<br/>
+     * The next major version can be assumed because the next major version can only ever
+     * talk to the highest minor of the previous major, so any features added before that
+     * will always exist on the cluster.<br/>
+     * It can be assumed for serverless because we never go backwards on serverless,
+     * so once a feature is in the serverless environment it will always be there.
+     */
+    public static boolean featuresCanBeAssumedForNode(DiscoveryNode node, Settings settings) {
+        return DiscoveryNode.isStateless(settings) || node.getVersion().major == Version.CURRENT.major + 1;
```

Comment:
Having to use "isStateless" seems to me like a failure of the "major version" being added to BuildVersion. 

Could we delegate this to BuildVersion, like have other things? Something like `isFutureMajorVersion()`, similar to `isFutureVersion()`? Then serverless can always return true, and isStateless would no longer be needed here.

Reply:
Ah yes, that's a good idea. That would simplify it significantly here too

Reply:
Done